cylc/flow/xtriggers/workflow_state.py

```diff
@@ -14,18 +14,117 @@
+from cylc.flow.scripts.workflow_state import WorkflowPoller
+from cylc.flow.id import tokenise
+from cylc.flow.exceptions import WorkflowConfigError
+from cylc.flow.task_state import TASK_STATUS_SUCCEEDED
 
 
 def workflow_state(
+    workflow_task_id: str,
+    offset: Optional[str] = None,
+    flow_num: Optional[int] = 1,
+    is_output: bool = False,
+    is_message: bool = False,
+    alt_cylc_run_dir: Optional[str] = None,
+) -> Tuple[bool, Dict[str, Any]]:
+    """Connect to a workflow DB and check a task status or output.
+
+    If the status or output has been achieved, return {True, result}.
+
+    Arg:
+        workflow_task_id:
+            ID (workflow//point/task:selector) of the target task.
+        offset:
+            Offset from cycle point as an ISO8601 or integer duration,
+            e.g. PT1H (1 hour) or P1 (1 integer cycle)
+        flow_num:
+            Flow number of the target task.
+        alt_cylc_run_dir:
+            Alternate cylc-run directory, e.g. for another user.
+
+    Returns:
+        tuple: (satisfied, result)
+
+        satisfied:
+            True if ``satisfied`` else ``False``.
+        result:
+            Dict {workflow_id, task_id, task_selector, flow_num}.
+
+    """
+    poller = WorkflowPoller(
+        workflow_task_id,
+        offset,
+        flow_num,
+        alt_cylc_run_dir,
+        TASK_STATUS_SUCCEEDED,
+        is_output, is_message,
+        old_format=False,
+        condition=workflow_task_id,
+        max_polls=1,  # (for xtriggers the scheduler does the polling)
+        interval=0,  # irrelevant for 1 poll
+        args=[]
+    )
+    if asyncio.run(poller.poll()):
+        return (
+            True,
+            {
+                "workflow_id": poller.workflow_id,
+                "task_id": f"{poller.cycle}/{poller.task}",
+                "task_selector": poller.selector,
+                "flow_num": poller.flow_num
+            }
+        )
+    else:
+        return (
+            False,
+            {}
+        )
+
+
+def validate(args: Dict[str, Any]):
+    """Validate workflow_state xtrigger function args.
+
+    Arguments:
+        workflow_task_id:
+            full workflow//cycle/task[:selector]
+        offset:
+            must be a valid status
+        flow_num:
+            must be an integer
+        alt_cylc_run_dir:
+            must be a valid path
+
+    """
+    try:
+        tokens = tokenise(args["workflow_task_id"])
+    except KeyError:
+        raise WorkflowConfigError(
+            # TODO better message
+            "Full ID needed: workflow//cycle/task[:selector].")
+
+    if any(
+        tokens[token] is None
+        for token in ("workflow", "cycle", "task")
+    ):
+        raise WorkflowConfigError(
+            "Full ID needed: workflow//cycle/task[:selector].")
+
+    try:
+        int(args["flow_num"])
+    except ValueError:
+        raise WorkflowConfigError("flow_num must be an integer.")
```

Comment:
But `"3.141492"` (str) would. I think we used to parse all xtrigger args from the config file as strings, and *convert* them to type like this, in which case `isinstance` would fail. `int("1") = 1`. But we seem to be doing better than that now (tested), so probably should do it your way... changing it

Reply:
(Now tweaked further to handle the None case, after changing the default to None).